src/utils/twoFactor.js

```diff
@@ -0,0 +1,265 @@
+            accountId,
+            method
+        });
+    }
+
+    async reInitTwoFactor(accountId, method) {
+        // clear any previous requests in localStorage (for verifyTwoFactor)
+        setRequest({})
+        return this.sendRequest(accountId, method)
+    }
+
+    async resend(accountId, method) {
+        if (!accountId) accountId = this.wallet.accountId
+        if (!method) method = await this.get2faMethod()
+        const requestData = getRequest()
+        let { requestId, data } = requestData
+        if (!requestId && requestId !== 0) {
+            requestId = -1
+        }
+        return this.sendRequest(accountId, method, requestId, data)
+    }
+
+    // requestId is optional, if included the server will try to confirm requestId
+    async verifyTwoFactor(accountId, securityCode) {
+        if (this.wallet.tempTwoFactorAccount) {
+            accountId = this.wallet.tempTwoFactorAccount.accountId
+        }
+        const requestData = getRequest()
+        let { requestId } = requestData
+        if (!requestId && requestId !== 0) {
+            requestId = -1
+        }
+        // try to get a accountId for the request
+        if (!accountId) accountId = requestData.accountId || this.wallet.accountId
+        if (!accountId) {
+            return
+        }
+        return await this.wallet.postSignedJson('/2fa/verify', {
+            accountId,
+            securityCode,
+            requestId
+        });
+    }
+
+    async sendMoney(account, receiver_id, amount) {
+        const request = {
+            receiver_id,
+            actions: [{ type: 'Transfer', amount }]
+        }
+        return await this.request(account, request)
+    }
+
+    async request(account, request) {
+        if (this.wallet.tempTwoFactorAccount) account = this.wallet.tempTwoFactorAccount
+        const { accountId } = account
+        const contract = new nearApiJs.Contract(account, account.accountId, {
+            viewMethods: VIEW_METHODS,
+            changeMethods: METHOD_NAMES_LAK,
+            sender: account.accountId
+        });
+        await deleteUnconfirmedRequests(contract)
+        const request_id = await getNextRequestId(contract)
+        try {
+            await contract.add_request_and_confirm({ request })
+        } catch (e) {
+            throw(e)
+        }
+        const request_id_after = await getNextRequestId(contract)
+        if (request_id_after > request_id) {
+            const data = { request_id, request }
+            const method = await this.get2faMethod()
+            return await this.sendRequest(accountId, method, request_id, data)
+        }
+    }
+
+    async addKey(account, publicKey, contractId, fullAccess = false) {
+        const {accountId} = account
+        const accessKeys = await this.wallet.getAccessKeys(accountId)
+        if (accessKeys.find((ak) => ak.public_key.toString() === publicKey)) {
+            throw new Error(`key already exists for account ${accountId}`)
```

Comment:
for non-2FA we consider it a success (though probably should also check if desired limitations (`contractId`, etc) match